pkg/meta/redis.go

```diff
@@ -120,45 +133,64 @@ func newRedisMeta(driver, addr string, conf *Config) (Meta, error) {
 				fopt.SentinelAddrs[i] = net.JoinHostPort(h, port)
 			}
 		}
-		fopt.Username = opt.Username
-		fopt.Password = opt.Password
-		if fopt.Password == "" {
-			fopt.Password = os.Getenv("REDIS_PASSWORD")
-		}
-		if fopt.Password == "" {
-			fopt.Password = os.Getenv("META_PASSWORD")
-		}
 		fopt.SentinelPassword = os.Getenv("SENTINEL_PASSWORD")
 		fopt.DB = opt.DB
+		fopt.Username = opt.Username
+		fopt.Password = opt.Password
 		fopt.TLSConfig = opt.TLSConfig
-		fopt.MaxRetries = conf.Retries
-		fopt.MinRetryBackoff = query.duration("min-retry-backoff", time.Millisecond*20)
-		fopt.MaxRetryBackoff = query.duration("max-retry-backoff", time.Second*10)
-		fopt.ReadTimeout = query.duration("read-timeout", time.Second*30)
-		fopt.WriteTimeout = query.duration("write-timeout", time.Second*5)
+		fopt.MaxRetries = opt.MaxRetries
+		fopt.MinRetryBackoff = opt.MinRetryBackoff
+		fopt.MaxRetryBackoff = opt.MaxRetryBackoff
+		fopt.ReadTimeout = opt.ReadTimeout
+		fopt.WriteTimeout = opt.WriteTimeout
 		if conf.ReadOnly {
 			// NOTE: RouteByLatency and RouteRandomly are not supported since they require cluster client
 			fopt.SlaveOnly = query.Get("route-read") == "replica"
 		}
 		rdb = redis.NewFailoverClient(&fopt)
 	} else {
-		if opt.Password == "" {
-			opt.Password = os.Getenv("REDIS_PASSWORD")
-		}
-		if opt.Password == "" {
-			opt.Password = os.Getenv("META_PASSWORD")
+		if !strings.Contains(opt.Addr, ",") {
+			c := redis.NewClient(opt)
+			info, err := c.ClusterInfo(Background).Result()
+			if err != nil && strings.Contains(err.Error(), "cluster mode") || err == nil && strings.Contains(info, "cluster_state:") {
+				logger.Infof("redis %s is in cluster mode", opt.Addr)
+			} else {
+				rdb = c
+			}
+		}
+		if rdb == nil {
+			var copt redis.ClusterOptions
+			copt.Addrs = strings.Split(opt.Addr, ",")
+			copt.MaxRedirects = 1
+			copt.Username = opt.Username
+			copt.Password = opt.Password
+			copt.TLSConfig = opt.TLSConfig
+			copt.MaxRetries = opt.MaxRetries
+			copt.MinRetryBackoff = opt.MinRetryBackoff
+			copt.MaxRetryBackoff = opt.MaxRetryBackoff
+			copt.ReadTimeout = opt.ReadTimeout
+			copt.WriteTimeout = opt.WriteTimeout
+			if conf.ReadOnly {
```

Comment:
It's probably not required for cluster mode.

Reply:
the replication is still asynchronous